Add car service functionality

Sir-Beard opened the pull request:
added a lot of files, changes

Sir-Beard committed:
Taxi service. Task2. Commit #4.
Small edit in ConnectionUtil.java class.
Commit-push into add-spring branch.
21:57 08.06.2023

Sir-Beard committed:
Revert "Taxi service. Task2. Commit #4."

This reverts commit afa5a7a106f54963d479f1733edd8c91e3770944.

Sir-Beard committed:
Taxi service. Task3. Commit #1.
a lot of changes
12:05 22.06.2023

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+            statement.setLong(1, car.getManufacturer().getId());
+            statement.setString(2, car.getModel());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
```
у цьому методі потрібно також обробити ситуацію, коли при створенні машини в ній вже буде сет водіїв всередині. Тобто необхідно додати потенційні зв'язки з водіями при створенні як і у методі апдейт

zaitsev-serhei replied:
і в методі створення після того як додаси функціонал оновлення потенційних зв'язків з водіями, також потрібно зробити транзакції

Sir-Beard replied:
Транзакції додані.
Зв'язок коли сет водіїв додається до машини при її створенні - зроблено.

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+
+    private Set<Driver> getDrivers(ResultSet resultSet) throws SQLException {
+        Driver driver = new Driver();
+        final Set<Driver> drivers = new HashSet<>();
```
нащо тут файнал ?

Sir-Beard replied:
без final - Maven скаржиться що оголошення змінної (Set<Driver> drivers = new HashSet<>()) і її "наповнення" (drivers.add(driver);) знаходяться дуже далеко один від одної (більше ніж 3 строчки)

Sir-Beard replied:
final прибраний, рядок оголошення змінної та рядок її використання знаходяся на мінімальній відстані один від одного, так щоб maven не скаржився на те, що вони далеко знаходяться один від одного.

zaitsev-serhei commented:
А мені здається, чи в структурі проекта дійсно дві папки Interfaces, а також немає інтерфейсу для Manufacturer?

zaitsev-serhei commented:
Maven окрім того, не пройшов перевірку на сайті

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+
+    @Override
+    public Car update(Car car) {
+        String queryUpdate
```
В методі апдейт потрібно додати функціонал транзакцій. **Не зі Spring**-а

Sir-Beard replied:
Функціонал транзацій доданий, згідно з https://docs.oracle.com/javase/tutorial/jdbc/basics/transactions.html

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        car.setDeleted(isDeleted);
+        car.setManufacturer(getManuf(resultSet));
+        resultSet.getObject("Driver ID", Long.class);
+        if (!(resultSet.wasNull())) {
```
наскільки я бачу таке рішення також працює, однак Леонід може забракувати його )))

Sir-Beard replied:
if (!(resultSet.wasNull())) {
видалено через непотрібне використання. код був перероблений

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        driver.setName(name);
+        driver.setLicenseNumber(licenseNumber);
+        driver.setDeleted(isDeleted);
+        drivers.add(driver);
```
Зараз виходить, що ми завжди повертаємо сет лише з 1м водієм. А якщо в резалт сеті у нас буде 2 водія, чи більше? Тут явно не вистачає цикла або якогось опрацювання декількох водіїв. Чи я щось пропустив у коді виклику метода?

Sir-Beard replied:
був доданий додатковий функціонал у вигляді методу getAllDriversFromOneCar
який записує водіїв до Сету водіїв в класі Car.

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+
+    @Override
+    public Optional<Car> get(Long id) {
+        String queryGet = "SELECT c.id AS 'Car ID', c.manufacturer_id AS 'Manufacturer ID Linked', c.model AS 'Car model', c.is_deleted, "
```
Протестить с 2+ водителями

Sir-Beard replied:
протестував, виводить більше одного водія, якщо у машини такі є.

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        return car;
+    }
+
+    private Manufacturer getManuf(ResultSet resultSet) throws SQLException {
```
acturers

Sir-Beard replied:
😄 👍

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        return drivers;
+    }
+
+    private void removeCarDriver(Car car, Connection connection) {
```
```suggestion
    private void removeCarDrivers(Car car, Connection connection) {
```

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        }
+    }
+
+    private void addCarDriver(Car car, Connection connection) {
```
```suggestion
    private void addCarDrivers(Car car, Connection connection) {
```

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,262 @@
+        ) {
+            statement.setLong(1, car.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
```
remove catch

Sir-Beard replied:
catch removed from removeCarDrivers and addCarDrivers methods

Sir-Beard committed:
Taxi service. Task3. Commit 2.
1.Main class became more structured for better testing.
2.Big changes affected CarDaoImpl.java class:
sql queries were rewritten to include Manufacturer update and Driver update. Added methods to interact with Driver and Manufacturer classes and to write data into db properly.
3.Added transactions to dao classes (methods with direct impact on DB: like create/update).
timestamp: 20:44 06.07.2023

Sir-Beard committed:
Taxi service. Task3. Commit 3.(without push)
CarDaoImpl.java - removed old comments, added TODO comments.
pom.xml - ?
timestamp: 15:51 07.07.2023

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,278 @@
+
+    @Override
+    public Car create(Car car) {
+        String queryCreate = "INSERT INTO cars (manufacturer_id, model) VALUES (?, ?)";
```
потрібно винести запис нової машини в окремий приватний метод, як і обговорювали раніше

Sir-Beard replied:
Зроблений окремий приватний метод для запису даних при створенні об'єкту Машина.
Те саме і для update методу.
P.s. також були оновлені DriverDaoImpl та ManufacturerDaoImpl класи.

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,278 @@
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create car: "
+                    + car, e);
+        }
```
потрібно додати rollback для транзакцій

Sir-Beard replied:
rollback додані до усіх DaoImpl класів

zaitsev-serhei commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,278 @@
+                        = connection.prepareStatement(queryUpdate)
+        ) {
+            connection.setAutoCommit(false);
+            statement.setLong(1, car.getManufacturer().getId()); // винести в окремий метод
```
як і в методі create потрібно винести в окремий приватний метод

Sir-Beard replied:
Запис у бд винесено в окремий приватний метод.

Sir-Beard committed:
Taxi service. Task3. Commit 3.
CarDaoImpl.java class modified, "create" and "update" methods reworked, rollback added.
timestamp: 16:41 14.07.2023

Sir-Beard committed:
Taxi service. Task3. Commit 4.
Commiting pom.xml file with <spring.version>6.0.9</spring.version> from local, instead of <spring.version>5.2.21.RELEASE</spring.version> from github.
timestamp: 17:16 14.07.2023

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+        try {
+            connection.setAutoCommit(false);
+            createCarInDb(car, connection);
+            connection.commit();
```
```suggestion
            createCarInDb(car, connection);
            addCarDriver(car, connection);
            connection.commit();
```

Sir-Beard replied:
Пересунув addCarDrivers назад в create метод.
Пересунув removeCarDrivers, addCarDrivers назад в update метод.

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+        resultSet.next();
+        Long insertedId = resultSet.getObject(1, Long.class);
+        car.setId(insertedId);
+        addCarDriver(car, connection);
```
```suggestion
```

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+        return car;
+    }
+
+    private void createCarInDb(Car car, Connection connection) throws SQLException {
```
Стараемся не делать void методы. Тем более для создания традиционно принято возвращать созданный объект
```suggestion
    private Car createCarInDb(Car car, Connection connection) throws SQLException {
```

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+            try {
+                connection.rollback();
+            } catch (SQLException ex) {
+                throw new RuntimeException("Rollback is not successful", ex); // TODO: прибрати це, і додати логгер?
```
TODO: прибрати це, і додати логгер?
- да

Sir-Beard replied:
коментар прибрав.
логер додав в наступній гілці під назвою add-logger-log4j2

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+
+    private void createCarInDb(Car car, Connection connection) throws SQLException {
+        String queryCreate = "INSERT INTO cars (manufacturer_id, model) VALUES (?, ?)";
+        PreparedStatement statement = connection.prepareStatement(queryCreate, PreparedStatement.RETURN_GENERATED_KEYS);
```
создавай PreparedStatement в блоке TWR

Sir-Beard replied:
Переписав insertCarDataToDb і modifyCarDataInDb щоб PreparedStatement був у try-with-resources.

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,301 @@
+        statement.setString(2, car.getModel());
+        statement.setLong(3, car.getId());
+        int rowsUpdated = statement.executeUpdate();
+        removeCarDriver(car, connection);
+        addCarDriver(car, connection);
```
вынеси эти методы наружу и вызывай их по очереди в update методе.
Таким образом, когда мы читаем код update метода, то мы видим алгоритм, последовательность действий. И если нужно - переходим из update метода на уровень ниже.

Sir-Beard replied:
done.

Rommelua commented on src/main/java/com/taxi/dao/DriverDaoImpl.java:
```diff
@@ -105,11 +106,13 @@ public Driver update(Driver driver) {
             statement.setLong(3, driver.getId());
             int rowsUpdated = statement.executeUpdate();
             if (rowsUpdated > 0) {
+                connection.commit();
```
ролбек?

Sir-Beard replied:
додано в наступних комітах

Rommelua commented on src/main/java/com/taxi/dao/ManufacturerDaoImpl.java:
```diff
@@ -43,6 +43,7 @@ public Manufacturer create(Manufacturer manufacturer) {
                 Long insertedId = resultSet.getObject(1, Long.class);
                 manufacturer.setId(insertedId);
             }
+            connection.commit();
```
ролбек?

Sir-Beard replied:
додано в наступних комітах

Sir-Beard committed:
Taxi service.
edit pom.xml spring-version

Sir-Beard committed:
Taxi service. Task3. Commit 6.
1. Renamed some methods in CarDaoImpl.java
2. SOLID principle for DriverDaoImpl.java and ManufacturerDaoImpl.java classes, refactoring "create" and "update" methods to have Single Responsibility Principle.
3. Added "rollback" to DriverDaoImpl.java and ManufacturerDaoImpl.java classes.
4. Reworked some "void" methods to have type and return type, in classes: CarDaoImpl.java, DriverDaoImpl.java, ManufacturerDaoImpl.java.
timestamp: 23:54 18.07.2023

Sir-Beard committed:
Taxi service. Task3. Commit 7.

Виправлений код згідно коментарям на github.

timestamp: 21:56 19.07.2023

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,312 @@
+                    + car, e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
+                connection.close();
```
ставить setAutoCommit(true); перед закрытием нет смысла

Sir-Beard replied:
прибрав

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,312 @@
+                    + car, e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
```
ставить setAutoCommit(true); перед закрытием нет смысла

Sir-Beard replied:
прибрав

Rommelua commented on src/main/java/com/taxi/dao/CarDaoImpl.java:
```diff
@@ -0,0 +1,312 @@
+        return car;
+    }
+
+    private Car modifyCarDataInDb(Car car, Connection connection) throws SQLException {
```
Стараемся придерживаться единой терминологии
```suggestion
    private Car updateCarDataInDb(Car car, Connection connection) throws SQLException {
```

Sir-Beard replied:
переписав create та update

Rommelua commented on src/main/java/com/taxi/dao/DriverDaoImpl.java:
```diff
@@ -26,25 +25,78 @@ public DriverDaoImpl(ConnectionUtil connectionUtil) {
                     + driver, e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
```
ставить setAutoCommit(true); перед закрытием нет смысла

Sir-Beard replied:
прибрав

Rommelua commented on src/main/java/com/taxi/dao/DriverDaoImpl.java:
```diff
@@ -26,25 +25,78 @@ public DriverDaoImpl(ConnectionUtil connectionUtil) {
+                    + driver, e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
```
ставить setAutoCommit(true); перед закрытием нет смысла

Sir-Beard replied:
прибрав

Rommelua commented on src/main/java/com/taxi/dao/DriverDaoImpl.java:
```diff
@@ -26,25 +25,78 @@ public DriverDaoImpl(ConnectionUtil connectionUtil) {
+        return driver;
+    }
+
+    private Driver modifyDriverDataInDb(Driver driver, Connection connection) throws SQLException {
```
```suggestion
    private Driver updateDriverDataInDb(Driver driver, Connection connection) throws SQLException {
```

Sir-Beard replied:
переписав create та update

Rommelua commented on src/main/java/com/taxi/dao/ManufacturerDaoImpl.java:
```diff
@@ -26,26 +25,79 @@ public ManufacturerDaoImpl(ConnectionUtil connectionUtil) {
                     + manufacturer, e);
+        } finally {
+            try {
+                connection.setAutoCommit(true);
```
ставить setAutoCommit(true); перед закрытием нет смысла

Sir-Beard replied:
прибрав

Sir-Beard committed:
Taxi service. Task3. Commit 8.
Correction DaoImpl classes, according to comments to previous commit.
timestamp: 00:08 21.07.2023
